e2e/actions/actions.go

```diff
@@ -65,7 +65,8 @@ func actionRun(t *testing.T) {
 	for _, tt := range tests {
 		e2e.RunSingularity(
 			t,
-			tt.name,
+			testenv.CmdPath,
```

Comment:
We should get rid of `testenv.CmdPath`, because it's a superficial argument and instead we could append `filepath.Join(buildcfg.LIBEXECDIR, "singularity/bin")` to `PATH` in `Run` function: https://github.com/sylabs/singularity/blob/master/e2e/suite.go#L38 and just restore `singularity` value in `RunSingularity`

Reply:
@cclerget 

I think the argument is that we may not run the installed singularity binary each time. I think that makes sense to make as an input to testing.

Reply:
After slack discussion, this argument is not required and instead we could set `PATH` conditionally to handle case for a different binary locations (pro release tests ...) :

```
path := os.Getenv("PATH")
if path == "" {
    path = "/usr/local/bin:/usr/local/sbin:/usr/sbin:/usr/bin:/sbin:/bin"
}
// to customize path for both singularity and external binaries
// used in e2e tests
e2ePath := os.Getenv("E2E_PREPEND_PATH")
if e2ePath == "" {
    e2ePath = filepath.Join(buildcfg.LIBEXECDIR, "singularity/bin")
}
os.Setenv("PATH", fmt.Sprintf("%s:%s", e2ePath, path))
```

Reply:
No, messing with PATH to get the "right" value is the wrong approach here.

Keep in mind that this is a test and the thing being tested is singularity: the path to the singularity binary under test is an input to the test. That binary under test should be capable of finding its components (confguration file, starter binaries, etc) on its own. How the top level test harness obtains the location of this binary should not be a concern of the test.

I suspect the problem you are trying to work around of is the fact that we are passing a *testing.T argument to the test functions, when we probably should pass something like a *e2e.T which embeds (or includes a field that is) a *testing.T. This *e2e.T would contain the information that we want to pass from the top level test harness down to the tests. If we want to do discovery of test functions, it's probably a good idea to define a single type for the functions that implement tests (just like the testing package does).

All that said, what I'm trying to fix in this PR is the incorrect assumption that the singularity binary under test is the first singularity program found in PATH because that's messing with what I actually want to do.

Reply:
There is confusion, the snippet above is intended to be in `e2e/suite.go` not in `RunSingularity`, then you could just restore value to `"singularity"` in `RunSingularity` and `LookPath` in `exec.Command` will does its job by finding the right singularity binary based on `PATH` initialization above and would avoid to pass this argument everytime. Now if you really need to pass a custom singularity binary, you could add a function like `WithBinary(path)` to `RunSingularity`

Reply:
No, I understood the suggestion correctly. I still say it's going the wrong way about it: it is up to the test harness to locate the binary that is going to be tested. singularity, for better or worse, is built in such a way that it will look in specific paths for the binaries it needs. From the point of view of the E2E tests, you must install singularity (as in `make install`) and from there singularity itself will locate its stuff. The E2E harness is using buildcfg to locate the install location of the singularity binary, so it's testing the binary that matches the current build. That does not need to be that way, but it is that way right now. By looking for singularity in the PATH, as the current code does, it is making a strong and wrong assumption that whatever singularity binary is found that way will match the singularity binary from the current build. In general, that's false.

Reply:
My bad, I edited code snippet above to use the right `buildcfg` path value.

I'm just pointing that we can get rid of this useless argument for `RunSingularity` ... I suggested to initialize `PATH` like above because we can put a custom path first in `PATH` where `LookPath` during call to `exec.Command("singularity" ...)` in `RunSingularity` will consider the first singularity binary found as the one to execute for e2e tests, this doesn't differ of what was done previously with `cmdPath := filepath.Join(buildcfg.BINDIR, "singularity")`.

If you are not happy with `PATH`, this is not a problem, just back to the original issue and the thing I pointed out in my first comment is: why passing singularity binary path as an argument of `RunSingularity` ? which always has the same value and which could involve more work in case of a refactoring, when `exec.Command` in `RunSingularity` could use an initialized variable directly like `exec.Command(e2e.SingularityPath, ...)` or `PATH` or whatever ...

And like I said, if a test require to use another singularity binary for some reasons but I can't figure out a use case for that, there is the possibility to use something like `WithBinary(path)` as a `SingularityCmdOp` argument for `RunSingularity`.

Reply:
If I understand, @cclerget is trying to make the following point:

Adding the `cmdPath` argument to the `RunSingularity()` function violates the principle of DRY (Don't repeat yourself).  There are ~50 instances of the `RunSingularity()` function that need to be altered in this PR after the argument is added.  And in all 50 calls, the variable is going to be the same right?  Seems tedious and error prone.  So why not just add something like this _into the `RunSingularity()` function itself._
```
cmdPath := filepath.Join(buildcfg.BINDIR, "singularity")
```
And then use it there once instead of setting it outside the function and passing it in 50 times?  (Or you could also just make it a constant in the `e2e` package and import it into the `RunSingularity()` function or whatever.  The point remains the same.)

Is that basically what you are trying to say @cclerget or am I totally misunderstanding your point?

Reply:
Yes that's the point @GodloveD